### DISCUSSION TITLE
分享2.5 exercises解法

### DISCUSSION THREAD

**Discussion by jzh18** [mlc-ai/mlc-zh#35] [Show and tell]
https://github.com/JZh18/mlc-ai-exercises/blob/main/2_5_Exercises_for_TensorIR.ipynb

**Comment by hbsun2113** [comment]
请问为什么要把parallel提前？我尝试把parallel放到提示位置，就会报错：
![image](https://user-images.githubusercontent.com/27994442/177684465-d7dda529-8cfc-435d-bc9d-1ebdda6e23f1.png)

我的code: https://colab.research.google.com/gist/hbsun2113/ff3a7d96a1f83904f1d9aeab61b9e836/2-5-exercises-for-tensorir-ipynb.ipynb

**Reply by un-knight** [reply]
我也是，感觉有点迷

**Reply by hbsun2113** [reply]
@Hzfengsy 您能帮助解释下原因吗？

**Reply by Hzfengsy** [reply]
parallel需要在decompose reduction之前操作，具体原因在报错信息中有写。应该是注释位置有误

**Reply by hbsun2113** [reply]
请问为什么“parallel需要在decompose reduction之前操作”？
下图是报错信息，但是我没有看出具体违背了哪个condition，@hzfengsy 您能帮忙指出吗？
![I6O_U} DV2FVF3`UNUKFRPH](https://user-images.githubusercontent.com/27994442/178869549-de1ddb1c-9a9b-437d-9d32-47d0e1adeca1.jpg)

**Reply by Hzfengsy** [reply]
`Y_update` 不是reduction block，违反了第一条（没有init statement， 即`with T.init`）

**Reply by Sanzo00** [reply]
> `Y_update` 不是reduction block，违反了第一条（没有init statement， 即`with T.init`）

我想确认下，`Y_update`不是compute blokc的原因是因为违反了原则3吗？（输入和输出重叠）

`3) No overlap between the buffers the block reads and writes`

**Reply by yzh119** [reply]
> 我想确认下，Y_update不是compute blokc的原因是因为违反了原则3吗？（输入和输出重叠）

是complete block不是compute block。
是的，complete block一般是不包含reduction的计算，读写区域没有重叠。

**Reply by Sanzo00** [reply]
明白了，感谢回复。

**Comment by GeLee-Q** [comment]
https://github.com/GeLee-Q/TVM_Practice/blob/master/TensorIR_Practice.ipynb

我写的时候也发现了这个迷惑点 hhhhhh
